app/src/main/java/ch/epfl/sdp/appart/database/firestoreservicehelpers/FirestoreAdHelper.java

```diff
@@ -0,0 +1,278 @@
+    private void checkAdUpload(CompletableFuture<Void> result, Ad ad, DocumentReference adRef,
+                               DocumentReference cardRef, StorageReference imagesRef,
+                               List<String> imagesRefsList) {
+        adRef.set(extractAdInfo(ad)).addOnCompleteListener(
+                task -> {
+                    cleanUpIfFailed(task.isSuccessful(), result, adRef, cardRef, imagesRef);
+                    result.complete(null);
+                });
+        // TODO is this the right place to do this?
+        setPhotosReferencesForAd(imagesRefsList, adRef, cardRef, imagesRef);
+    }
+
+    /**
+     * Checks whether the upload of card information to Firestore completed successfully. If it
+     * didn't, cleans up and completes exceptionally.
+     */
+    private void checkCardUpload(CompletableFuture<Void> result, Ad ad, DocumentReference adRef,
+                                 DocumentReference cardRef, StorageReference imagesRef,
+                                 String firstImageRef) {
+        Card c = new Card(cardRef.getId(), adRef.getId(), ad.getAdvertiserId(), ad.getCity(),
+                ad.getPrice(), firstImageRef, ad.hasVRTour());
+        cardHelper.putCard(c, cardRef).thenAccept(successful -> {
+            cleanUpIfFailed(successful, result, adRef, cardRef, imagesRef);
+            result.complete(null);
+        });
+    }
+
+    /**
+     * Checks that every part of the ad upload completed successfully, completes exceptionally if at
+     * least one Future completed exceptionally.
+     */
+    private void finalizeAdUpload(CompletableFuture<String> result, CompletableFuture<Void> imagesResult,
+                                  CompletableFuture<Void> adResult, CompletableFuture<Void> cardResult,
+                                  DocumentReference adRef) {
+        CompletableFuture<Void> allOf = CompletableFuture.allOf(adResult, cardResult);
+        allOf.thenAccept(res -> result.complete(adRef.getId()));
+        allOf.exceptionally(e -> {
+            result.completeExceptionally(e);
+            return null;
+        });
+    }
+
+    /**
+     * If failed, deletes from firebase the given references and completes exceptionally
+     */
+    private void cleanUpIfFailed(boolean taskSuccessful, CompletableFuture<Void> result, DocumentReference adRef,
+                                 DocumentReference cardRef, StorageReference imagesRef) {
+        if (!taskSuccessful) {
+            Log.d("Ad creation", "ad upload failed");
+            adRef.delete();
+            cardRef.delete();
+            imagesRef.listAll().addOnSuccessListener(listResult -> {
+                List<StorageReference> items = listResult.getItems();
+                for (StorageReference item : items) {
+                    item.delete();
+                }
+            }).addOnFailureListener(e -> Log.d("Ad upload", "Failed to cleanup after failed upload"));
+
+            // TODO create exception
```

Comment:
Left TODO.

Reply:
It's been weeks we should have started discussing about designing the way we deal with database errors, it's not up to this PR to define all the exceptions for database operations. The TODO is there as a reminder.